### PR TITLE
Utility functions for handling effects

### PR DESCRIPTION
**Pull request by arybczak**
Supersedes #230.

@michaelpj @ocharles does that work for you?

I decided to go with `handleWith` instead of `interpretWith` since the name `interpretWith` is too close to `interpret` and might be confusing.

Still reads nicely:

```haskell
myFunction arg `handleWith_` \case
  NotifyA i -> ...
  NotifyB s -> ...
```

and won't clash with `Control.Monad.Catch`.

### CODE REVIEW

**Comment by arybczak** [comment]
Well, actually maybe `interpretWith` is better? Because then I can also add `interposeWith` for a flipped `interpose`. Damn naming :sweat_smile:

**Comment by ocharles** [comment]
I think your comment is a good suggestion. Then we don't really introduce any new terms. So I'm +1 on `fooWith` meaning "flipped `foo`".

**Comment by michaelpj** [comment]
Yeah, that would have been what I would have gone for.

**Comment by arybczak** [comment]
Ok, I added all possible variants, it should be good to go.

File: effectful-core/src/Effectful/Dispatch/Dynamic.hs [hunk 1]
```diff
@@ -435,6 +464,19 @@ reinterpret runHandlerEs handler m = unsafeEff $ \es -> do
...
+-- @since 2.4.0.0
+reinterpretWith
+  :: DispatchOf e ~ Dynamic
+  => (Eff handlerEs a -> Eff es b)
+  -- ^ Introduction of effects encapsulated within the handler.
+  -> Eff (e : es) a
+  -> EffectHandler e handlerEs
+  -- ^ The effect handler.
+  -> Eff      es  b
+reinterpretWith runHandlerEs m handler = reinterpret runHandlerEs handler m
```
[the remark this answers is not in the document]
Comment (arybczak) [reply effectful-core/src/Effectful/Dispatch/Dynamic.hs:478]: In my experience `runHandlerEs` is usually a single internal runner like `runReader` or `runState` (maybe both?). In the event that it's an additional setup, like

```haskell
blah = reinterpret run $ \case
  ...
  where
    run m = do
      x <- setup1
      y <- setup2
      runState x $ runReader y m
```

`reinterpretWith` should actually help with this:

```haskell
blah = do
  x <- setup1
  y <- setup2
  reinterpretWith (runState x . runReader y) m $ \case
    ...
```

On the other hand, the handler is almost always multiline, so if you can't partially apply `reinterpret`, you need to bind the handler separately.

To not look very far, the code from https://github.com/haskell-effectful/effectful/discussions/229 would benefit from `reinterpretWith`.

Am I missing something? :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (michaelpj) [reply effectful-core/src/Effectful/Dispatch/Dynamic.hs:478]: No, that's convincing!

**Comment by michaelpj** [comment]
Thanks!

**Pull request merged**